hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/ServiceStateNode.java

```diff
@@ -0,0 +1,202 @@
+     * @param stateKey The state key. Cannot be null. It must be at least one character in length,
+     *     and must be comprised of spaces, numbers, and alphabetic characters in the ascii range.
+     * @param merkle The merkle node to set. Cannot be null.
+     */
+    public void put(@NonNull final String stateKey, @NonNull final MerkleNode merkle) {
+        validateStateKey(stateKey);
+        final int existingIndex = indexOf(stateKey);
+        final int index = existingIndex == -1 ? getNumberOfChildren() : existingIndex;
+        setChild(index, new StringLeaf(stateKey));
+        setChild(index + 1, merkle);
+    }
+
+    /**
+     * Removes the state associated with the given state key, if any. This method is idempotent.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    public void remove(@NonNull final String stateKey) {
+        Objects.requireNonNull(stateKey);
+        final int matchingIndex = indexOf(stateKey);
+        if (matchingIndex != -1) {
+            setChild(matchingIndex - 1, null);
+            setChild(matchingIndex, null);
+        }
+    }
+
+    @Override
+    public long getClassId() {
+        return CLASS_ID;
+    }
+
+    @Override
+    public int getVersion() {
+        return CURRENT_VERSION;
+    }
+
+    @Override
+    @NonNull
+    public ServiceStateNode copy() {
+        throwIfImmutable();
+        throwIfDestroyed();
+        setImmutable(true);
+        return new ServiceStateNode(this);
+    }
+
+    /**
+     * Get the index of the state associated with the given state key, or null if there is no such
+     * state.
+     *
+     * @param stateKey The state key, cannot be null
+     * @return The index of the associated state, or -1 if there is not any
+     */
+    private int indexOf(@NonNull final String stateKey) {
+        Objects.requireNonNull(stateKey);
+
+        // The first child of ServiceStateNode is a StringLeaf containing the serviceName. The
+        // subsequent children come in pairs -- the first is a StringLeaf with the stateKey,
+        // the next is the corresponding MerkleNode.
+        final var numChildren = getNumberOfChildren();
+        for (int i = 1; i < numChildren - 1; i += 2) {
+            final StringLeaf idNode = getChild(i);
+            if (idNode != null && stateKey.equals(idNode.getValue())) {
+                return i + 1;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Verifies the state key meets all the validation requirements.
+     *
+     * @param stateKey The state key
+     * @throws NullPointerException if the state key is null
+     * @throws IllegalArgumentException if any other validation criteria fails
+     */
+    private void validateStateKey(@NonNull final String stateKey) {
+        if (Objects.requireNonNull(stateKey).isEmpty()) {
+            throw new IllegalArgumentException("The state key must have characters");
+        }
+
+        for (int i = 0; i < stateKey.length(); i++) {
```

Comment:
A regular expression would be easier IMO.

Reply:
This class is now removed